src/Common/Core/Impl/Threading/BinaryAsyncLock.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Common.Core.Threading {
+    /// <summary>
+    /// BinaryAsyncLock is a helper primitive that can be used instead of SemaphoreSlim.WaitAsync + double-checked locking
+    /// After BinaryAsyncLock is created or reset, the first caller of <see cref="WaitAsync"/> will immediately get <see langword="false" />
+    /// All other callers will wait until <see cref="Release"/> is called and then will get <see langword="true" />
+    /// </summary>
+    public class BinaryAsyncLock {
```

Comment:
I was thinking of a property on the returned object instead of comparing it to bool directly, but either one would work; and since you'd always want to check when using this object, I guess the latter would make more sense. That would take `operator true` and `operator false`, right?